### PR TITLE
Replace GetModuleFileName with GetBinaryLocation (fix #2297)

### LINKED ISSUE #2297: PAL: Cleanup GetModuleFileName** and implement corresponding interface
This is a follow-up issue for #2296 

```
+// xplat-todo: Implement a corresponding solution for GetModuleFileNameW
+// and cleanup PAL. [ https://github.com/Microsoft/ChakraCore/pull/2288 should be merged first ]
+// GetModuleFileName* PAL is not reliable and forces us to explicitly double initialize PAL
+// with argc / argv....
```

~~Waiting for #2288~~

Move `GetBinaryLocation` (https://github.com/Microsoft/ChakraCore/blob/7541e228ef42a97895a88f46d5427d07bf2675ba/bin/ch/Helpers.cpp#L577) into PlatformAgnostic and use that instead of `GetModuleFileName*`

edit-2, Some more details:

- Create `SystemInfo.cpp` under `lib/Runtime/PlatformAgnostic/Platform/Common/`

```
#include "Common.h"
#include "ChakraPlatform.h"

namespace PlatformAgnostic
{
    void SystemInfo::GetBinaryLocation(........) {   .......  }
    ....
```

- After moving GetBinaryLocation under the new SystemInfo.cpp, we should add it into `lib/Runtime/PlatformAgnostic/CMakeLists.txt` (i.e. at line 4)

- Similarly add that source file into `lib/Runtime/PlatformAgnostic/Chakra.Runtime.PlatformAgnostic.vcxproj` (i.e. line 41)

- Finally add SystemInfo's GetBinaryLocation static method definition into `lib/Common/PlatformAgnostic/SystemInfo.h`

### PR DESCRIPTION
**Pull request by xiaoyinl**
This will fix the remaining task in Issue #2297: replace `GetModuleFileName` with `PlatformAgnostic::SystemInfo::GetBinaryLocation`.

This task is easy if the caller of `GetModuleFileName` can easily work with a UTF-8 string, like `TTDHostBuildCurrentExeDirectory`. But if the caller needs to pass the returned string to a Unicode version of Windows API (see #3178 for example), then it seems the UTF-8 string returned by `GetBinaryLocation` is not so suitable here. What to do in this case?

### CODE REVIEW

**Comment by xiaoyinl** [comment]
How about we implement a UTF-16 version of `GetBinaryLocation`, i.e.  `PlatformAgnostic::SystemInfo::GetBinaryLocationW` for WIN32 only? What do you think @obastemur?

File: bin/ch/Helpers.cpp [hunk 1]
```diff
@@ -7,30 +7,41 @@
...
+    while (exePath[i] != TTD_HOST_PATH_SEP[0] && i != 0)
     {
         --i;
     }
-
-    if(i * 3 > bufferLength)
+    if (i == 0)
     {
-        printf("Don't overflow path buffer during conversion");
+        printf("%s", exePath);  // If exePath doesn't contain path separator, it's an error message
```
Comment (obastemur) [inline bin/ch/Helpers.cpp]: Could be nice to show an error message here (along with the exePath)

**Comment by obastemur** [comment]
@xiaoyinl Thanks for this! overall looks good. Could you please take a look at the review comment and squash them all ?

File: bin/ch/Helpers.cpp (hunk 1 above)
Comment (xiaoyinl) [inline bin/ch/Helpers.cpp]: But based on what we did in #3112, `GetBinaryLocation` already returns an error message in `exePath`: `"GetBinaryLocation: GetModuleFileName has failed."`, etc. 

Or do you want me to add another error message along with that message?

Comment (obastemur) [inline bin/ch/Helpers.cpp]: Could be nice to see why we print the path there. i.e. `this xxx path is empty or not found etc..`

Comment (xiaoyinl) [inline bin/ch/Helpers.cpp]: I mean this line doesn't print a path. `exePath` itself is an error message.
The `GetBinaryLocation` function sets `exePath` to either an actual path or an error message. (such as [L52 of SystemInfo.h](https://github.com/xiaoyinl/ChakraCore/blob/eed45f6ae78e004502614afc34edefd5367a325a/lib/Common/PlatformAgnostic/SystemInfo.h#L52))
Here the `while` loop finds the index of the path separator ("\\" or "/") in `exePath`. If `i == 0`, the path separator doesn't exist, so `exePath` is not a path, but an error message.

Comment (obastemur) [inline bin/ch/Helpers.cpp]: You comment seems sufficient there.

**Comment by obastemur** [comment]
BTW; can we also replace the other usages under bin/ch ? and lib/Common/Core/ConfigParser.cpp?

**Comment by xiaoyinl** [comment]
@obastemur Not so easily. For example, `ConfigParser::SetOutputFile` in `lib/Common/Core/ConfigParser.cpp` needs `moduleName` to be an UTF-16 string because it later passes it to `GetEnvironmentVariableW`, but our GetBinaryLocation returns UTF-8.

File: bin/ch/Helpers.cpp (hunk 1 above)
Comment (xiaoyinl) [inline bin/ch/Helpers.cpp]: So does this line look good? Or do you mean to change the printf message to the comment?

**Comment by obastemur** [comment]
@xiaoyinl could it be done by overloading `PlatformAgnostic::SystemInfo::GetBinaryLocation` ?

**Comment by xiaoyinl** [comment]
@obastemur I will try it.

**Comment by xiaoyinl** [comment]
I have implemented `bool SystemInfo::GetBinaryLocation(char16 *path, const unsigned size)`. It needs to return `bool` because some callers check the return value of `GetModuleFileName`. On Linux, this function calls the original `GetBinaryLocation` to get the `path` and then uses `utf8::DecodeUnitsIntoAndNullTerminate` to convert the string from UTF-8 to UTF-16. For this reason, the original `GetBinaryLocation` has to return `bool` so that the overloaded one can propagate the result to its caller.

Does it look good to you, @obastemur?

**Comment by obastemur** [comment]
@xiaoyinl great effort! Let me check the handle part.

**Comment by obastemur** [comment]
@xiaoyinl have you tried with a Unicode path? 

Curious if we could add a test file (name containing Unicode chars)

**Comment by xiaoyinl** [comment]
@obastemur #3178 is an issue when the name contains Unicode chars. I didn't find other issues with Unicode path. I'm not sure how to add test cases for that.

**Comment by dilijev** [comment]
@obastemur @xiaoyinl any update here?

**Comment by obastemur** [comment]
Before fixing conflicts, wait for #4088

**Comment by dilijev** [comment]
Note: https://github.com/Microsoft/ChakraCore/pull/4088 was closed in favor of https://github.com/Microsoft/ChakraCore/pull/4141 which was merged. This PR should be unblocked now.

**Comment by obastemur** [comment]
@xiaoyinl could you please rebase?

**Comment by xiaoyinl** [comment]
@obastemur  Sorry for the delay. I have rebased the change. Is there anything else I need to do in this PR?

**Comment by obastemur** [comment]
@xiaoyinl Thanks for the effort! LGTM. I will merge.

File: bin/ch/Helpers.cpp [hunk 2]
```diff
@@ -7,30 +7,41 @@
 #include <sys/stat.h>
 
 //TODO: x-plat definitions
-#ifdef _WIN32
 #define MAX_URI_LENGTH 512
+#ifdef _WIN32
+#define TTD_MAX_FILE_LENGTH MAX_PATH
 #define TTD_HOST_PATH_SEP "\\"
+#else
+#define TTD_MAX_FILE_LENGTH MAX_URI_LENGTH
```
Comment (obastemur) [inline bin/ch/Helpers.cpp]: `MAX_URI_LENGTH` definition looks redundant.

File: bin/ch/Helpers.cpp [hunk 3]
```diff
@@ -7,30 +7,41 @@
...
+    while (exePath[i] != TTD_HOST_PATH_SEP[0] && i != 0)
     {
         --i;
     }
-
-    if(i * 3 > bufferLength)
+    if (i == 0)
     {
-        wprintf(_u("Don't overflow path buffer during conversion"));
+        wprintf(_u("Can't get current exe directory"));
```
Comment (obastemur) [inline bin/ch/Helpers.cpp]: `fprintf(stderr, ...` instead?

File: bin/ch/Helpers.cpp [hunk 4]
```diff
@@ -7,30 +7,41 @@
...
         exit(1);
     }
-    *pathLength = utf8::EncodeInto((LPUTF8)path, exePath, (charcount_t)(i + 1));
+    if (i + 2 > bufferLength)
+    {
+        printf("Don't overflow path buffer during copy");
```
Comment (obastemur) [inline bin/ch/Helpers.cpp]: `fprintf(stderr, ...` instead?

**Comment by obastemur** [comment]
@xiaoyinl I was about to merge but couple of tiny details bugged me. I do appreciate if you may look at them?

File: bin/ch/Helpers.cpp (hunk 4 above)
Comment (dilijev) [reply bin/ch/Helpers.cpp]: Also printf is not allowed anymore in ChakraCore because it will cause an assertion on Windows because printf can't handle our upgrade to UTF-8 output.

Comment (dilijev) [reply bin/ch/Helpers.cpp]: We prefer using `Output::Print`. At a minimum use `wprintf` and `fwprintf`

Comment (obastemur) [reply bin/ch/Helpers.cpp]: I don't get the part.. why we use wprintf once the string is not unicode? i.e. the string above? All the `_u(x)` etc. stuff is an unnecessary hassle.

Comment (dilijev) [reply bin/ch/Helpers.cpp]: It is probably okay because the string is ASCII, but as a general policy we don't use `printf`, and instead use `Output::Print` (which takes a unicode string).

Comment (dilijev) [reply bin/ch/Helpers.cpp]: (I think in `ch` we don't have access to the `Output` class so we use `wprintf` instead.)

Comment (xiaoyinl) [reply bin/ch/Helpers.cpp]: So use `fwprintf(stderr, ...);`?

Comment (dilijev) [reply bin/ch/Helpers.cpp]: Yeah, I think that would be best. @Penguinwizzard since you're making more changes in the area of stdlib output, what do you think?

Comment (Penguinwizzard) [reply bin/ch/Helpers.cpp]: Yeah, use fwprintf(stderr, ...) for error for now; the _u is a hassle, but I'm not entirely sure if mixing printf and wprintf works nicely in all cases (especially since setting up output encoding properly is already a bit nuanced).

Comment (xiaoyinl) [reply bin/ch/Helpers.cpp]: I have changed both to `fwprintf(stderr, ...);`

Comment (xiaoyinl) [reply bin/ch/Helpers.cpp]: @Penguinwizzard Do you want me to change other occurrences of `fprintf` and `wprintf` in this file to `fwprintf`?

Comment (dilijev) [reply bin/ch/Helpers.cpp]: @xiaoyinl up to you, but I'd say since it's out of scope of this change and it hasn't broken yet, if you want to make changes like that you could do it in a follow-up PR.

File: bin/ch/Helpers.cpp (hunk 2 above)
Comment (xiaoyinl) [reply bin/ch/Helpers.cpp]: `MAX_URI_LENGTH` is needed in functions `Helpers::CreateTTDDirectoryAsNeeded` (L393) and `Helpers::TTCreateStreamCallback` (L457).

Comment (xiaoyinl) [reply bin/ch/Helpers.cpp]: Or do you mean change this line to `#define TTD_MAX_FILE_LENGTH 512`?

File: bin/ch/Helpers.cpp (hunk 4 above)
Comment (xiaoyinl) [reply bin/ch/Helpers.cpp]: I see.

Comment (Penguinwizzard) [reply bin/ch/Helpers.cpp]: @xiaoyinl Thanks for offering, but I actually have a change in the works (but backlogged because of more important work) that does a lot of those changes.

File: bin/ch/Helpers.cpp (hunk 2 above)
Comment (obastemur) [reply bin/ch/Helpers.cpp]: Yes.
